pkg/metrics/providers/appdynamicscloud.go

```diff
@@ -0,0 +1,214 @@
+	tenantAddress        string
+	metricsQueryEndpoint string
+	clientSecretID       string
+	clientSecretKey      string
+
+	timeout time.Duration
+	client  *http.Client
+}
+
+// NewAppDynamicsCloudProvider takes a provider spec and the credentials map,
+// and returns a AppDynamicsCloud client ready to execute queries against the API
+func NewAppDynamicsCloudProvider(
+	provider flaggerv1.MetricTemplateProvider,
+	credentials map[string][]byte) (*AppDynamicsCloudProvider, error) {
+
+	address := provider.Address
+	if address == "" {
+		return nil, fmt.Errorf("appdynamics cloud endpoint url address is not set")
+	}
+
+	tid, err := getTenantID(address)
+	if tid == "" || err != nil {
+		return nil, fmt.Errorf("failed to retrieve tenant id based on tenant URL address: %s", address)
+	}
+
+	appdCloudProvider := AppDynamicsCloudProvider{
+		tenantID:             tid,
+		tenantAddress:        address,
+		metricsQueryEndpoint: address + metricsQueryPath,
+
+		timeout: 5 * time.Second,
+	}
+
+	if b, ok := credentials[clientSecretID]; ok {
+		appdCloudProvider.clientSecretID = string(b)
+	} else {
+		return nil, fmt.Errorf("appdynamics cloud credentials does not contain %s", clientSecretID)
+	}
+
+	if b, ok := credentials[clientSecretKey]; ok {
+		appdCloudProvider.clientSecretKey = string(b)
+	} else {
+		return nil, fmt.Errorf("appdynamics cloud credentials does not contain %s", clientSecretKey)
+	}
+
+	return &appdCloudProvider, nil
+
+}
+
+// RunQuery executes the appdynamics cloud query against AppDynamicsCloudProvider
+// metricsQueryEndpoint and returns the result as float64
+func (p *AppDynamicsCloudProvider) RunQuery(query string) (float64, error) {
+	if p.client == nil {
+		if _, err := p.IsOnline(); err != nil {
+			return 0, fmt.Errorf("failed to login to query endpoint: %w", err)
+		}
+	}
+
+	jsonQuery, err := json.Marshal(map[string]string{
+		"query": query,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("failed to marshal query: %w", err)
+	}
+
+	// output to flagger container runtime log
+	// fmt.Print("appdynamicscloud metric query:", string(jsonQuery))
+	p.client.Timeout = p.timeout
+	resp, err := p.client.Post(p.metricsQueryEndpoint, "application/json", bytes.NewBuffer(jsonQuery))
+	if err != nil {
+		return 0, fmt.Errorf("failed to get query response: %w", err)
+	}
+
+	body, _ := io.ReadAll(resp.Body)
+	resp.Body.Close()
+
+	// we want to extract just a single float value from the result, hence no need to
+	// un-marshalling the entire struct using Appdynamics Cloud Query API
+	var anyValue []map[string]any
+	if err := json.Unmarshal(body, &anyValue); err != nil {
+		return 0, fmt.Errorf("failed to un-marshaling result: %s.\n error: %w", string(body), err)
+	}
+	if len(anyValue) < 1 {
+		return 0, fmt.Errorf("invalid response for query: %s: %w", query, ErrNoValuesFound)
+	}
+	// actual result (non-meta data) is in the data element of the last item
+	// of the json array
+	data := anyValue[len(anyValue)-1]["data"].([]any)
```

Comment:
Hi @aryan9600 ,

Thank you for reviewing and the questions. AppDynamics Cloud Query response is a big model that covers metrics, logs, traces and spans. I am hoping to reduce the code complexity since we only retrieves a single floating value from the upstream. As long as the query submitted to the upstream is for a single metrics value, the result is either empty DataResultChunk or an array of {timestamp, numbers} pair with the last item being the value that we wanted. Rather than including the entire package of the API data model and un-marshaling all the meta data, what I try to do is to put enough error checking around each type cast so that we can safely return the numeric value at the end. I just realized that I need to error check the last type cast to float64 as well. Will it be good enough once that is added? 

Thanks,

Charles

Reply:
Hi @aryan9600,

I added error checking for type cast of the last item of the array to float. Love to hear your thoughts. 

Thanks,

Charles